### ISSUE TITLE
Need to attach private network through annotations

### ISSUE THREAD

**Issue by karthick-kk** [UpCloudLtd/uks-instructions#43]
# Why?
We have a private DNS which connects to the private Loadbalancer DNS(Private hostnames)
The frontend is attached to the private network. This is to send communication entirely over internal network(10.x.x.x) and not to expose our traffic over private internet.
# Issue:
The setup we have is working good until we try to scale up/down the Kubernetes node group. Whenever the node count is updated, underlying LB reconfigure itself to include/update the backed nodes. This is an expected action, however the LB reconfiguration also removes the attached private network from the frontend, causing entire traffic to be halted. This is an unexpected behavior and we would like to have the private network remains attached.
# Expectation: 
This is our current annotation to configure our LB to work in TCP mode.
 
```
service.beta.kubernetes.io/upcloud-load-balancer-config: |
      {
          "plan": "development",
          "frontends": [
              {
                  "name": "http",
                  "mode": "tcp",
                  "port": 80
              },
              {
                  "name": "https",
                  "mode": "tcp",
                  "port": 443
              }
          ]
      }
      
```
  We would like to include private network attachment as part of the annotation, to avoid network detachment behavior during loadbalancer reconfiguration.
  
```
   service.beta.kubernetes.io/upcloud-load-balancer-config: |
      {
          "plan": "development",
          "frontends": [
              {
                  "name": "http",
                  "mode": "tcp",
                  "port": 80
              },
              {
                  "name": "https",
                  "mode": "tcp",
                  "port": 443,
                  "networks": [
                    {
                      "name": "public-IPv4"
                    },
                    {
                      "name": "private-IPv4"
                    }
                    ]
              }
          ]
      }
```

**Comment by i-prudnikov** [comment]
Hello, thank you for sharing this.
Could you please clarify, if it is an option for you not to attach additional network but to use the existing network that both - workers and LB uses?

If not, how about creating additional network and route to existing one (that workers and LB already uses)? This allow to reach LB frontend through private network and be fully independent on LB configuration. In this case you should:
- use private DNS name of the LB (available on hub UI)
- create router upfront and attach to existing network then point out to use this network when creating k8s cluster (let's name it k8s network)
- create new network that will be used to access to k8s network, and attach it to existing router, make sure that route entries are injected by DHCP (setting while attaching networks, IIRC)

Please let me know what you think?

**Comment by karthick-kk** [comment]
The usecase is to utilize the same network that LB and workers runs on. 
 We need to have both the Public Network and Private Network attached to the LB. This is because some services connect externally and some services are completely private. For instance, appA(public) and appB(private) has this traffic flow
 
 appA.domain.com -> LB(94.x.x.x) -> nginx ingress -> serviceA
 
 appB.domain.com -> LB(10.0.1.6) -> nginx ingress -> serviceB
 
 currently, appA.domain.com is being pointed to lb-0a1f403da6f247b4b329455707113b5a-1.upcloudlb.com
 and appB.domain.com is being pointed to lb-0a1f403da6f247b4b329455707113b5a-2.upcloudlb.com
 
 Both the apps are utilizing the same LB and there is an ingress service that routes the traffic based on incoming domain name. 
 
 And we have multiple other apps that connect via public interface and multiple apps that connect via private. Since the private nw gets detached all the services routed through private nw of LB becomes inaccessible.

**Comment by i-prudnikov** [comment]
Ok, so your option is to have either routed network, or use the same  network that k8s cluster runs (workers and LBs). I see no problem to implement that, without manual attaching network to LB.

Option 1, same network, private traffic:
appB.domain.com -> (you should be able to reach LB network from client side)-> LB(10.0.1.6) -> nginx ingress -> serviceB

Option 2, routed network:
appB.domain.com -> ( routed network)->DHCP-Route-> LB(10.0.1.6) -> nginx ingress -> serviceB

Public traffic will work the same no matter which option you follow.

So the question from where you actually going to use DNS name appB.domain.com , technically, I mean that this DNS names should be configured on some apps/clients that run on some servers in particular network. If it is upcloud managed network, then you either use the same network that your worker and LB uses (option 1) or separate network (that routed to k8s network) option 2. 

Side comment: the network where LB runs is not allowed to be configured intentionally, as this can break connectivity (imagine you select network that is not routed to workers network) between workers and LB, and LB will not be able to serve it's role - delivering traffic from outside to k8s cluster.

**Comment by karthick-kk** [comment]
Option 1 is the exact scenario we are seeking to achieve. And it is infact working out of the box, with public and private network(network where LB and workers reside) in normal conditions. The issue arise when there is nodegroup reconfiguration due to cluster autoscaler or manual node scaling. The LB tries to reconfigure and remove/detach the private network from the frontend. 

Earlier, we also had an issue with the LB mode resetting itself to HTTP when we manally updated the configuration to TCP. The support team provided us LB annotations to lock the mode to TCP and it worked perfectly till now.

We are looking forward to get similar solution in the form of annotations as stated in the issue so that the nw is locked from automatic detachment during cluster reconfiguration.


We don't host any dns servers, the dns is configured at digitalocean but pointed to private hostname of the LB.

On the side comment, I have to agree if there are multiple networks, but I don't see anything else in the dropdown list other than "Private IPV4", and I hope it's always preconfigured to the network where LB and workers are in, therefore this should be perfectly fine for our usecase.

**Comment by i-prudnikov** [comment]
Looks like this is kind of misunderstanding.
You don't need to attach manually any network to frontend, it is already use the one where LB runs. So there is nothing to configure AFAIU.  Or I didn't get your use case?

**Comment by karthick-kk** [comment]
Exactly. I might missed few details. 
As it's the LB configuration is working, but it's losing out one it's configuration, i.e the private-IPv4 nw from preconfigured frontend. I'm not sure it's  a bug but we would like to be an workaround atleast in the form of annotations. Or if it's fixed in the upcloud backend, we don't need this feature altogether.

**Comment by i-prudnikov** [comment]
Ok, am I understood it right that when LB scaled, (new member added/removed from/to backend) the private frontend loose it's attachment to private network and you are not able to to reach private LB interface? If this is the case, then it is definitely a bug.

**Comment by karthick-kk** [comment]
@i-prudnikov That's correct. Original configuration(in frontend) is lost when new nodes are added/removed to/from backend. 

Please do note that the LB in discussion was created by kubernetes controller(ingress service type set to LoadBalancer), we reconfigure it again from the upcloud console to change its frontend mode from HTTP to TCP and certificates detached. Now with annotations added, we no longed have to change the mode from the console as the config is persisted now. However, we can't persist the original private network in attached state. Hope that clarified :-)

**Comment by i-prudnikov** [comment]
Sorry for that many questions, now I see what's the problem.
Could you please also share the LB configuration your use with service annotations for us to reproduce?

**Comment by karthick-kk** [comment]
Perfectly fine. I'm attaching the ingress service with upcloud LB annotations added. Also attached is the LB configuration as pulled from API call. 
Please let me know if you require additional information any.
[lbconfig.json.txt](https://github.com/UpCloudLtd/uks-instructions/files/12795248/lbconfig.json.txt)
[ingress-nginx-controller-svc_with-annotations.yaml.txt](https://github.com/UpCloudLtd/uks-instructions/files/12795249/ingress-nginx-controller-svc_with-annotations.yaml.txt)

**Comment by i-prudnikov** [comment]
Ok, I do confirm that there is a bug. Will address that soon.

**Comment by i-prudnikov** [comment]
@karthick-kk ,so this issue is addressed in new CCM version 0.0.14, that is delivered to your cluster. Default behaviour is to attach frontend to both interfaces - private and public. The docs for CCM will be updates soon. Please close the issue if everything is ok on your side.

**Comment by thevilledev** [comment]
Hey @karthick-kk! As mentioned, this has been fixed in our Cloud Controller Manager a while ago. We've also added additional documentation about customising LB frontends in [ccm/networks.md](https://github.com/UpCloudLtd/uks-instructions/blob/main/ccm/networks.md) so please have a look.

I'll proceed to close off this issue but feel free to comment if there's still something missing.
